### PR TITLE
Non-unicode primitive support and major Kanji database update

### PR DESCRIPTION
**Pull request by mjuhanne**
Add capability to use primitives that have no Unicode character representation. They are instead referenced with [primitive_name] tags in primitives and primitive_of lists as well as Heisig stories and comments.
The tags are then converted into image links when constructing their visual representation.

Included are .svg files for the non-Unicode primitives gathered from [https://github.com/cyphar/heisig-rtk-index](https://github.com/cyphar/heisig-rtk-index) repository. I have included also few extra ones that the repository was missing. UPDATE: These are now in the cursive/ directory which will be used later. More uniform look is achieved with a new set of primitive images created from the default font.

Added Stroke order graphs for the new non-Unicode primitives as well as bunch of missing ones for Unicode characters.   Attempt to load the SVGs are first tried from the new kanjivg-supplementary/ directory and if that fails, fall back to the original kanjivg-repository. This way we can keep the kanjivg repository separate and avoid copying files there if graphs are modified or new graphs are added.

Major update to kanji.db, trying to replicate Heisig's keywords and primitives as accurately as possible:

- Add all the missing non-Unicode character primitives
- Fix many mix-ups between primitives ('good luck' vs 'lidded crock', 'chop-seal' vs 'stamps' among others)
- separate 'hand' and 'fingers' primitives for clarity.
- primitives for 'flowers', 'city walls' and 'pinnacle' actually reference the more common radicals instead of weird archaic ones
- Fix many erroneous primitive references (such as 'water' being used instead of 'rice grains')
- Fix many Heisig stories and comments. Also add italic and bold text when referencing keywords

For existing decks a manual injection of these new primitives is not needed ('Refresh All' will scan for missing primitives and  a confirmation pop-up dialog will be shown).

Changes to the original kanji database (as of 7/2023) are merged from kanji-ext.tsv files. Included is the updated kanji.db. There are 6 different batches of updates present in this PR. Log of changes can be found in corresponding db_merge_log_X.md files in Markdown format (the log and tsv files are now removed for cleaner addon directory, but can be found in the individual commits below if needed)

### CODE REVIEW

File: addon/fonts.py [hunk 1]
```diff
@@ -23,6 +23,14 @@ def get_col_name(idx):
 def get_col_path(idx):
     return util.col_media_path(get_col_name(idx))
 
+def assure_primitive_col_media():
+    primitive_dir = util.addon_path("primitives")
+    primitive_files = os.listdir(primitive_dir)
+    for f in primitive_files:
+        src_path = util.addon_path("primitives", f)
+        dest_path =  util.col_media_path('_primitive_' + f)
+        shutil.copy(src_path, dest_path)
```
Comment (MinmoTech) [inline addon/fonts.py:32]: I currently get this error on the latest version:

```
Error
An error occurred. Please start Anki while holding down the shift key, which will temporarily disable the add-ons you have installed.
If the issue only occurs when add-ons are enabled, please use the Tools > Add-ons menu item to disable some add-ons and restart Anki, repeating until you discover the add-on that is causing the problem.
When you've discovered the add-on that is causing the problem, please report the issue to the add-on author.
Debug info:
Anki 2.1.66 (70506aeb) Python 3.9.15 Qt 6.5.0 PyQt 6.5.0
Platform: Linux-6.4.12-zen1-1-zen-x86_64-with-glibc2.35
Flags: frz=True ao=True sv=3
Add-ons, last update check: 2023-09-01 12:04:11
Add-ons possibly involved: ⁨migaku_kanji_addon⁩

Caught exception:
Traceback (most recent call last):
  File "aqt.progress", line 118, in handler
  File "aqt.main", line 217, in on_window_init
  File "aqt.main", line 316, in setupProfile
  File "aqt.main", line 495, in loadProfile
  File "_aqt.hooks", line 3977, in __call__
  File "/home/julius/.var/app/net.ankiweb.Anki/data/Anki2/addons21/migaku_kanji_addon/__init__.py", line 126, in on_profile_open
    CardType.upsert_all_models()
  File "/home/julius/.var/app/net.ankiweb.Anki/data/Anki2/addons21/migaku_kanji_addon/card_type.py", line 207, in upsert_all_models
    cls.assure_global_col_media()
  File "/home/julius/.var/app/net.ankiweb.Anki/data/Anki2/addons21/migaku_kanji_addon/card_type.py", line 198, in assure_global_col_media
    fonts.assure_col_media()
  File "/home/julius/.var/app/net.ankiweb.Anki/data/Anki2/addons21/migaku_kanji_addon/fonts.py", line 120, in assure_col_media
    assure_primitive_col_media()
  File "/home/julius/.var/app/net.ankiweb.Anki/data/Anki2/addons21/migaku_kanji_addon/fonts.py", line 32, in assure_primitive_col_media
    shutil.copy(src_path, dest_path)
  File "shutil", line 427, in copy
  File "shutil", line 264, in copyfile
IsADirectoryError: [Errno 21] Is a directory: '/home/julius/.var/app/net.ankiweb.Anki/data/Anki2/addons21/migaku_kanji_addon/primitives/cursive'
```

**Review by MinmoTech — CHANGES_REQUESTED** [review]

File: addon/fonts.py (hunk 1 above)
Comment (mjuhanne) [reply addon/fonts.py]: Ah, sorry, the cursive type images in the subdirectory that weren't going to be used yet in this version but I left them there for the next PR/commit.  Here's the patch that actually uses them and copies the subdirectory contents correctly.

**Review by MinmoTech — CHANGES_REQUESTED** [review]
When trying to do reviews (I.e. clicking on a deck and then "Study now") I get this error:
```
Error
An error occurred. Please start Anki while holding down the shift key, which will temporarily disable the add-ons you have installed.
If the issue only occurs when add-ons are enabled, please use the Tools > Add-ons menu item to disable some add-ons and restart Anki, repeating until you discover the add-on that is causing the problem.
When you've discovered the add-on that is causing the problem, please report the issue to the add-on author.
Debug info:
Anki 2.1.66 (70506aeb) Python 3.9.15 Qt 6.5.0 PyQt 6.5.0
Platform: Linux-6.4.12-zen1-1-zen-x86_64-with-glibc2.35
Flags: frz=True ao=True sv=3
Add-ons, last update check: 2023-09-01 12:04:11
Add-ons possibly involved: ⁨migaku_kanji_addon⁩

Caught exception:
Traceback (most recent call last):
  File "aqt.webview", line 46, in cmd
  File "aqt.webview", line 153, in _onCmd
  File "aqt.webview", line 665, in _onBridgeCmd
  File "aqt.overview", line 92, in _linkHandler
  File "aqt.main", line 712, in moveToState
  File "_aqt.hooks", line 4853, in __call__
  File "/home/julius/.var/app/net.ankiweb.Anki/data/Anki2/addons21/migaku_kanji_addon/reviewer.py", line 59, in learn_ahead_refresh_on_review_start
    aqt.mw.migaku_kanji_db.refresh_learn_ahead()
  File "/home/julius/.var/app/net.ankiweb.Anki/data/Anki2/addons21/migaku_kanji_addon/kanji.py", line 442, in refresh_learn_ahead
    self.make_cards_from_characters(ct, new, None)
  File "/home/julius/.var/app/net.ankiweb.Anki/data/Anki2/addons21/migaku_kanji_addon/kanji.py", line 618, in make_cards_from_characters
    aqt.mw.requireReset()
  File "aqt.main", line 838, in requireReset
  File "aqt.main", line 823, in reset
  File "aqt.main", line 775, in _synthesize_op_did_execute_from_reset
  File "_aqt.hooks", line 3716, in __call__
  File "aqt.main", line 783, in on_operation_did_execute
  File "aqt.reviewer", line 214, in op_executed
  File "aqt.reviewer", line 186, in refresh_if_needed
  File "aqt.reviewer", line 248, in nextCard
  File "aqt.reviewer", line 381, in _showQuestion
  File "aqt.reviewer", line 297, in _run_state_mutation_hook
AttributeError: 'Reviewer' object has no attribute '_state_mutation_js'
```

**Comment by mjuhanne** [comment]
> When trying to do reviews (I.e. clicking on a deck and then "Study now") I get this error:
> 
> ```
> Error
> An error occurred. Please start Anki while holding down the shift key, which will temporarily disable the add-ons you have installed.
> If the issue only occurs when add-ons are enabled, please use the Tools > Add-ons menu item to disable some add-ons and restart Anki, repeating until you discover the add-on that is causing the problem.
> When you've discovered the add-on that is causing the problem, please report the issue to the add-on author.
> Debug info:
> Anki 2.1.66 (70506aeb) Python 3.9.15 Qt 6.5.0 PyQt 6.5.0
> Platform: Linux-6.4.12-zen1-1-zen-x86_64-with-glibc2.35
> Flags: frz=True ao=True sv=3
> Add-ons, last update check: 2023-09-01 12:04:11
> Add-ons possibly involved: ⁨migaku_kanji_addon⁩
> 
> Caught exception:
> Traceback (most recent call last):
>   File "aqt.webview", line 46, in cmd
>   File "aqt.webview", line 153, in _onCmd
>   File "aqt.webview", line 665, in _onBridgeCmd
>   File "aqt.overview", line 92, in _linkHandler
>   File "aqt.main", line 712, in moveToState
>   File "_aqt.hooks", line 4853, in __call__
>   File "/home/julius/.var/app/net.ankiweb.Anki/data/Anki2/addons21/migaku_kanji_addon/reviewer.py", line 59, in learn_ahead_refresh_on_review_start
>     aqt.mw.migaku_kanji_db.refresh_learn_ahead()
>   File "/home/julius/.var/app/net.ankiweb.Anki/data/Anki2/addons21/migaku_kanji_addon/kanji.py", line 442, in refresh_learn_ahead
>     self.make_cards_from_characters(ct, new, None)
>   File "/home/julius/.var/app/net.ankiweb.Anki/data/Anki2/addons21/migaku_kanji_addon/kanji.py", line 618, in make_cards_from_characters
>     aqt.mw.requireReset()
>   File "aqt.main", line 838, in requireReset
>   File "aqt.main", line 823, in reset
>   File "aqt.main", line 775, in _synthesize_op_did_execute_from_reset
>   File "_aqt.hooks", line 3716, in __call__
>   File "aqt.main", line 783, in on_operation_did_execute
>   File "aqt.reviewer", line 214, in op_executed
>   File "aqt.reviewer", line 186, in refresh_if_needed
>   File "aqt.reviewer", line 248, in nextCard
>   File "aqt.reviewer", line 381, in _showQuestion
>   File "aqt.reviewer", line 297, in _run_state_mutation_hook
> AttributeError: 'Reviewer' object has no attribute '_state_mutation_js'
> ```

Damn it, looks like another bug in Anki. This may be partly because the addon uses legacy code which they haven't tested thoroughly. That code branch seems to run only with V3 scheduler so that's why I haven't come across that bug yet.

**Comment by mjuhanne** [comment]
The legacy GUI reset code seems to be broken in new Anki. I've removed temporarily calls to requireReset() and maybeReset() (the latter actually does nothing in the current Anki version).  Downside is that the GUI may not up to date after adding new characters (a minor nuisance).

Better option would be to port the add-on to use the new CollectionOp system in Anki which takes care of GUI updates and checkpoints (which also doesn't seem to work currently in with new Anki).   But this should be a topic for another PR, I think

**Comment by mjuhanne** [comment]
Fixed one more bug that caused color of user images to be inverted in dark mode (inversion is to be applied only to non-Unicode primitive images but in this case the CSS scope was too wide)

**Comment by calculuschild** [comment]
Testing this new branch, I'm unable to generate missing primitives automatically again (clicking "Kanji -> Refresh Cards").

For example, this character in particular: 亟 has "snare" 丂 as a primitive. The card 亟 itself has no story, but it does list the correct primitives. However when I refresh cards, no new cards are generated; that "snare" primitive is still missing.  I can add the primitive card manually though; it exists in the database.

**Comment by calculuschild** [comment]
I went ahead and deleted 50 cards from my deck. Clicking refresh fails to re-build any of them. They aren't anywhere in my deck now and the addon doesn't seem to recognize the are missing.

**Comment by ganqqwerty** [comment]
Great PR! To me, it is the last missing piece of the puzzle, after which KanjiGOD can really be used as a main kanji-learning tool. Is there any way I can help with the PR, what issue still needs to be resolved?

**Comment by calculuschild** [comment]
I've been using this branch for a while now and it's really nice. Would be great to see it completed!

**Review by hgiesel — APPROVED** [review]
I think by now we can surely say that this works. Any tests of mine were also successful. LGTM @MinmoTech

**Comment by ganqqwerty** [comment]
I installed it on Anki ⁨2.1.65 (aa9a734f)⁩, MacOS, so far I don't see any problems. I was waiting for this feature for a long time!

**Review by MinmoTech — APPROVED** [review]

**Pull request merged**